### ISSUE TITLE
Add prototype as a known property to classes

### ISSUE THREAD

**Issue by benmccann** [rollup/rollup#3587]
### Feature Use Case

I'd like to be able to set fields on the `.prototype` of a class and static members on a class. However, doing so stops them from being tree shaken out if the class is never read.

I'm a maintainer of the popular [Chart.js library](https://github.com/chartjs/Chart.js). We use rollup and are trying to convert our codebase to be compatible with tree shaking

### Feature Proposal

 It would be ideal if setting primitive values on a class definition would not prevent it from being tree shaken.

Here's a [repl example](https://rollupjs.org/repl/?version=2.10.8&shareable=JTdCJTIybW9kdWxlcyUyMiUzQSU1QiU3QiUyMm5hbWUlMjIlM0ElMjJtYWluLmpzJTIyJTJDJTIyY29kZSUyMiUzQSUyMmltcG9ydCUyMCU3QnRlc3QlN0QlMjBmcm9tJTIwJy4lMkZtb2RlbCclM0IlNUNuJTVDbnRlc3QoKSUzQiUyMiUyQyUyMmlzRW50cnklMjIlM0F0cnVlJTdEJTJDJTdCJTIybmFtZSUyMiUzQSUyMm1vZGVsLmpzJTIyJTJDJTIyY29kZSUyMiUzQSUyMmV4cG9ydCUyMGNsYXNzJTIwVGVzdCUyMCU3QiU1Q24lMjAlMjAlMjAlMjB0ZXN0KCklMjAlN0IlNUNuJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwY29uc29sZS5sb2coJ3Rlc3QnKSUzQiU1Q24lMjAlMjAlMjAlMjAlN0QlNUNuJTdEJTVDbiU1Q25UZXN0LmlkJTIwJTNEJTIwJ0EnJTNCJTVDbiU1Q25leHBvcnQlMjBjbGFzcyUyMFRlc3QyJTIwJTdCJTVDbiUyMCUyMCUyMCUyMHRlc3QoKSUyMCU3QiU1Q24lMjAlMjAlMjAlMjAlMjAlMjAlMjAlMjBjb25zb2xlLmxvZygndGVzdDInKSUzQiU1Q24lMjAlMjAlMjAlMjAlN0QlNUNuJTdEJTVDbiU1Q25UZXN0Mi5wcm90b3R5cGUuaWQlMjAlM0QlMjAnQSclM0IlNUNuJTVDbmV4cG9ydCUyMGNsYXNzJTIwVGVzdDMlMjAlN0IlNUNuJTIwJTIwJTIwJTIwdGVzdCgpJTIwJTdCJTVDbiUyMCUyMCUyMCUyMCUyMCUyMCUyMCUyMGNvbnNvbGUubG9nKCd0ZXN0MycpJTNCJTVDbiUyMCUyMCUyMCUyMCU3RCU1Q24lN0QlNUNuJTVDblRlc3QzJTIwJTNEJTIwJTdCJTdEJTNCJTVDblRlc3QzLmlkJTIwJTNEJTIwJ0EnJTNCJTVDbiU1Q24lNUNuZXhwb3J0JTIwY2xhc3MlMjBUZXN0NCUyMCU3QiU1Q24lMjAlMjAlMjAlMjB0ZXN0KCklMjAlN0IlNUNuJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwY29uc29sZS5sb2coJ3Rlc3Q0JyklM0IlNUNuJTIwJTIwJTIwJTIwJTdEJTVDbiU3RCU1Q24lNUNuT2JqZWN0LmRlZmluZVByb3BlcnR5KFRlc3Q0JTJDJTIwJ2lkJyUyQyUyMCU3QnZhbHVlJTNBJTIwJ0EnJTdEKSUzQiU1Q24lNUNuZXhwb3J0JTIwZnVuY3Rpb24lMjB0ZXN0KCklMjAlN0IlNUNuJTIwJTIwJTIwJTIwY29uc29sZS5sb2coJ3Rlc3RHJyklM0IlNUNuJTdEJTVDbiUyMiUyQyUyMmlzRW50cnklMjIlM0FmYWxzZSU3RCU1RCUyQyUyMm9wdGlvbnMlMjIlM0ElN0IlMjJmb3JtYXQlMjIlM0ElMjJlcyUyMiUyQyUyMm5hbWUlMjIlM0ElMjJteUJ1bmRsZSUyMiUyQyUyMmFtZCUyMiUzQSU3QiUyMmlkJTIyJTNBJTIyJTIyJTdEJTJDJTIyZ2xvYmFscyUyMiUzQSU3QiU3RCU3RCUyQyUyMmV4YW1wbGUlMjIlM0FudWxsJTdE)

**Comment by kzc** [comment]
Try using pure annotated IIFEs:

https://rollupjs.org/repl/?version=2.10.8&shareable=JTdCJTIybW9kdWxlcyUyMiUzQSU1QiU3QiUyMm5hbWUlMjIlM0ElMjJtYWluLmpzJTIyJTJDJTIyY29kZSUyMiUzQSUyMmltcG9ydCUyMCU3QnRlc3QlMkMlMjBUZXN0JTJDJTIwVGVzdDIlMkMlMjBUZXN0MyUyQyUyMFRlc3Q0JTdEJTIwZnJvbSUyMCcuJTJGbW9kZWwnJTNCJTVDbiU1Q250ZXN0KCklM0IlMjIlMkMlMjJpc0VudHJ5JTIyJTNBdHJ1ZSU3RCUyQyU3QiUyMm5hbWUlMjIlM0ElMjJtb2RlbC5qcyUyMiUyQyUyMmNvZGUlMjIlM0ElMjJleHBvcnQlMjBjbGFzcyUyMFRlc3QlMjAlN0IlNUNuJTIwJTIwdGVzdCgpJTIwJTdCJTIwY29uc29sZS5sb2coJ3Rlc3QnKSUzQiUyMCU3RCU1Q24lN0QlNUNuVGVzdC5pZCUyMCUzRCUyMCdBJyUzQiU1Q24lNUNuZXhwb3J0JTIwY29uc3QlMjBUZXN0MiUyMCUzRCUyMCUyRiolNDBfX1BVUkVfXyolMkYoKCklMjAlM0QlM0UlMjAlN0IlNUNuJTVDdGNsYXNzJTIwVGVzdDIlMjAlN0IlNUNuJTIwJTIwJTIwJTIwdGVzdCgpJTIwJTdCJTIwY29uc29sZS5sb2coJ3Rlc3QyJyklM0IlMjAlN0QlNUNuJTIwJTIwJTdEJTVDbiUyMCUyMFRlc3QyLnByb3RvdHlwZS5pZCUyMCUzRCUyMCdBJyUzQiU1Q24lNUN0cmV0dXJuJTIwVGVzdDIlM0IlNUNuJTdEKSgpJTNCJTVDbiU1Q25leHBvcnQlMjBjb25zdCUyMFRlc3QzJTIwJTNEJTIwJTJGKiU0MF9fUFVSRV9fKiUyRigoKSUyMCUzRCUzRSUyMCU3QiU1Q24lNUN0Y2xhc3MlMjBUZXN0MyUyMCU3QiU1Q24lMjAlMjAlMjAlMjB0ZXN0KCklMjAlN0IlMjBjb25zb2xlLmxvZygndGVzdDMnKSUzQiUyMCU3RCU1Q24lMjAlMjAlN0QlNUNuJTIwJTIwVGVzdDMlMjAlM0QlMjAlN0IlN0QlM0IlNUNuJTIwJTIwVGVzdDMuaWQlMjAlM0QlMjAnQSclM0IlNUNuJTIwJTIwcmV0dXJuJTIwVGVzdDMlM0IlNUNuJTdEKSgpJTNCJTVDbiU1Q25leHBvcnQlMjBjb25zdCUyMFRlc3Q0JTIwJTNEJTIwJTJGKiU0MF9fUFVSRV9fKiUyRigoKSUyMCUzRCUzRSUyMCU3QiU1Q24lNUN0Y2xhc3MlMjBUZXN0NCUyMCU3QiU1Q24lMjAlMjAlMjAlMjB0ZXN0KCklMjAlN0IlMjBjb25zb2xlLmxvZygndGVzdDQnKSUzQiUyMCU3RCU1Q24lMjAlMjAlN0QlNUNuJTIwJTIwT2JqZWN0LmRlZmluZVByb3BlcnR5KFRlc3Q0JTJDJTIwJ2lkJyUyQyUyMCU3QnZhbHVlJTNBJTIwJ0EnJTdEKSUzQiU1Q24lNUN0cmV0dXJuJTIwVGVzdDQlM0IlNUNuJTdEKSgpJTNCJTVDbiU1Q25leHBvcnQlMjBmdW5jdGlvbiUyMHRlc3QoKSUyMCU3QiU1Q24lMjAlMjBjb25zb2xlLmxvZygndGVzdEcnKSUzQiU1Q24lN0QlMjIlMkMlMjJpc0VudHJ5JTIyJTNBZmFsc2UlN0QlNUQlMkMlMjJvcHRpb25zJTIyJTNBJTdCJTIyZm9ybWF0JTIyJTNBJTIyZXMlMjIlMkMlMjJuYW1lJTIyJTNBJTIybXlCdW5kbGUlMjIlMkMlMjJhbWQlMjIlM0ElN0IlMjJpZCUyMiUzQSUyMiUyMiU3RCUyQyUyMmdsb2JhbHMlMjIlM0ElN0IlN0QlN0QlMkMlMjJleGFtcGxlJTIyJTNBbnVsbCU3RA==

**Comment by lukastaegert** [comment]
@benmccann As you probably saw, adding static properties directly already works if you do not overwrite your class like this:

```js
export class Test3 {
    test() {
        console.log('test3');
    }
}

// Here you overwrite your class with an empty object, which deoptimizes
// Rollup value tracking
Test3 = {}; 
Test3.id = 'A';
```

https://rollupjs.org/repl/?version=2.10.8&shareable=JTdCJTIybW9kdWxlcyUyMiUzQSU1QiU3QiUyMm5hbWUlMjIlM0ElMjJtYWluLmpzJTIyJTJDJTIyY29kZSUyMiUzQSUyMmltcG9ydCUyMCU3QnRlc3QlN0QlMjBmcm9tJTIwJy4lMkZtb2RlbCclM0IlNUNuJTVDbnRlc3QoKSUzQiUyMiUyQyUyMmlzRW50cnklMjIlM0F0cnVlJTdEJTJDJTdCJTIybmFtZSUyMiUzQSUyMm1vZGVsLmpzJTIyJTJDJTIyY29kZSUyMiUzQSUyMmV4cG9ydCUyMGNsYXNzJTIwVGVzdCUyMCU3QiU1Q24lMjAlMjAlMjAlMjB0ZXN0KCklMjAlN0IlNUNuJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwY29uc29sZS5sb2coJ3Rlc3QnKSUzQiU1Q24lMjAlMjAlMjAlMjAlN0QlNUNuJTdEJTVDbiU1Q25UZXN0LmlkJTIwJTNEJTIwJ0EnJTNCJTVDbiU1Q25leHBvcnQlMjBmdW5jdGlvbiUyMHRlc3QoKSUyMCU3QiU1Q24lMjAlMjAlMjAlMjBjb25zb2xlLmxvZygndGVzdEcnKSUzQiU1Q24lN0QlNUNuJTIyJTJDJTIyaXNFbnRyeSUyMiUzQWZhbHNlJTdEJTVEJTJDJTIyb3B0aW9ucyUyMiUzQSU3QiUyMmZvcm1hdCUyMiUzQSUyMmVzJTIyJTJDJTIybmFtZSUyMiUzQSUyMm15QnVuZGxlJTIyJTJDJTIyYW1kJTIyJTNBJTdCJTIyaWQlMjIlM0ElMjIlMjIlN0QlMkMlMjJnbG9iYWxzJTIyJTNBJTdCJTdEJTdEJTJDJTIyZXhhbXBsZSUyMiUzQW51bGwlN0Q=

`defineProperty` unfortunately is not recognized as a way to add properties yet and is therefore deoptimized at the moment.

Otherwise, we should probably add `prototype` as a known property to classes just as we do for functions, this could be achieved with reasonable effort:

https://rollupjs.org/repl/?version=2.10.8&shareable=JTdCJTIybW9kdWxlcyUyMiUzQSU1QiU3QiUyMm5hbWUlMjIlM0ElMjJtYWluLmpzJTIyJTJDJTIyY29kZSUyMiUzQSUyMmltcG9ydCUyMCU3QnRlc3QlN0QlMjBmcm9tJTIwJy4lMkZtb2RlbCclM0IlNUNuJTVDbnRlc3QoKSUzQiUyMiUyQyUyMmlzRW50cnklMjIlM0F0cnVlJTdEJTJDJTdCJTIybmFtZSUyMiUzQSUyMm1vZGVsLmpzJTIyJTJDJTIyY29kZSUyMiUzQSUyMmV4cG9ydCUyMGZ1bmN0aW9uJTIwVGVzdDIoKSUyMCU3QiU3RCU1Q24lNUNuVGVzdDIucHJvdG90eXBlLnRlc3QlMjAlM0QlMjBmdW5jdGlvbigpJTIwJTdCJTVDbiUyMCUyMGNvbnNvbGUubG9nKCd0ZXN0MicpJTNCJTVDbiU3RCUzQiU1Q24lNUNuVGVzdDIucHJvdG90eXBlLmlkJTIwJTNEJTIwJ0EnJTNCJTVDbiU1Q25leHBvcnQlMjBmdW5jdGlvbiUyMHRlc3QoKSUyMCU3QiU1Q24lMjAlMjAlMjAlMjBjb25zb2xlLmxvZygndGVzdEcnKSUzQiU1Q24lN0QlNUNuJTIyJTJDJTIyaXNFbnRyeSUyMiUzQWZhbHNlJTdEJTVEJTJDJTIyb3B0aW9ucyUyMiUzQSU3QiUyMmZvcm1hdCUyMiUzQSUyMmVzJTIyJTJDJTIybmFtZSUyMiUzQSUyMm15QnVuZGxlJTIyJTJDJTIyYW1kJTIyJTNBJTdCJTIyaWQlMjIlM0ElMjIlMjIlN0QlMkMlMjJnbG9iYWxzJTIyJTNBJTdCJTdEJTdEJTJDJTIyZXhhbXBsZSUyMiUzQW51bGwlN0Q=

**Comment by benmccann** [comment]
> we should probably add prototype as a known property to classes just as we do for functions,

That would be great!

I wasn't aware it was done for functions today, so thanks for pointing that out. @kzc I appreciate your tip as well, which might be what we try to do first.

**Comment by lukastaegert** [comment]
Fix at #3598
